node_manager.go

```diff
@@ -0,0 +1,991 @@
+/*
+ * File: node_manager.go
+ *
+ * Description: functions to handle nodes and node creation
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	clientv3 "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"golang.org/x/net/context"
+)
+
+//
+// Node definition
+//
+type Node struct {
+	HostName string
+	HostID   string
+}
+
+//! Update the TTL for a given reserved entry
+/*
+ * @param    LeaseID    id of a given etcd key entry
+ *
+ * @return   none
+ */
+func (inst *EtcdInstance) updateTTL(leaseID clientv3.LeaseID) {
+
+	// keep the entry alive...
+	_, err := inst.internal.KeepAliveOnce(context.TODO(), leaseID)
+
+	// if an error occurs, dump it to stdout
+	if err != nil {
+		stdlog(err.Error())
+		stdlog("updateTTL() --> could not update TTL")
+	}
+}
+
+//! Function to keep etcd key values alive
+/*
+ * @param    LeaseGrantResponse    lease response
+ *
+ * @return   none
+ */
+func (inst *EtcdInstance) keepKeyAlive(lease *clientv3.LeaseGrantResponse) {
+
+	// set a duration time based on the time-to-live
+	sleep_duration := time.Duration(lease.TTL / 2)
+
+	// infinite loop
+	for {
+
+		// update the time-to-live
+		inst.updateTTL(lease.ID)
+
+		// then go back to sleep
+		time.Sleep(time.Second * sleep_duration)
+	}
+}
+
+//! Prime a given node to make it ready for jobs
+/*
+ * @param    chan    channel for listening who is primed
+ *
+ * @return   none
+ */
+func (inst *EtcdInstance) primedLock(primedNotificationChan chan bool) {
+
+	// grant the lease
+	lease, err := inst.internal.Grant(context.TODO(), primedTTL)
+
+	// if an error occurs, print it out
+	if err != nil {
+		stdlog("primedLock(): failed to issue a TTL for the future primed lock: " + err.Error())
+		return
+	}
+
+	// setup a key-value store
+	kvc := clientv3.NewKV(inst.internal)
+
+	// grab the context
+	ctx, cancel := context.WithTimeout(context.Background(), etcdGracePeriod*time.Second)
```

Comment:
My original design was flawed on account of every improperly packaging files. Since this has been moved into the 'libetcd' package, the code can be improved and I have done so.

Your comments were rather helpful in this case given the haphazard nature of that function. Thank you very much 👍 

https://github.com/rbisewski/lclusterd/commit/811af0f9c6681a98db396d348fbf285dd3577be0